### ISSUE TITLE
[BUG] Metric plot slow for many data points

### ISSUE THREAD

**Issue by pimdh** [mlflow/mlflow#5779]
### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [X] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No, this concerns the UI.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04.3 LTS
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.25.1
- **Python version**: 3.8.10
- **Exact command to reproduce**: `mlflow ui --backend-store-uri "sqlite:///mlruns.db"`

### Describe the problem
Currently, in the MLFlow tracking UI metrics plot, when plotting around 100k points, the plot becomes unusably slow.

It appears like no simplification / decimation / subsampling strategy is used and all data points are simply rendered.

As I prefer to log metrics at all training steps, this is very problematic. Tensorboard, for example, subsamples the data points at the server-side, making plotting fast. This is configured via a [CLI argument](https://github.com/tensorflow/tensorboard/blob/348af4cf9564b8217750e8d84e0b2b89555be62c/tensorboard/data/server/cli.rs#L129).

The issue appears to lie in the rendering part, as the large JSON files containing all the metrics loads sufficiently fast.

Possible solutions:
- Subsample the data points server-side
- Subsample the data points client-side
- Use an adaptive method on the client side, such as https://github.com/predict-idlab/plotly-resampler 

### Tracking information
```
MLflow version: 1.25.1
Tracking URI: sqlite:///mlruns.db
```
```
mlflow ui --backend-store-uri "sqlite:///mlruns.db"
```

### Code to reproduce issue
Any plot with around 100k points.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by dbczumar** [comment]
Hi @pimdh , thank you for raising this! We would be very excited about integrating https://github.com/predict-idlab/plotly-resampler to improve plot scalability. I think it would be best to introduce a toggle that users can select to explicitly enable sampling, ensuring that they can still access the complete set of data (even if it's slow). We would be thrilled to review and provide guidance on a PR that implements this capability. Please let me know if you have any questions.

**Comment by pimdh** [comment]
Hi @dbczumar,
It seems like the library I mentioned is based on the Dash framework and thus appears difficult to integrate with the existing code-base. At the moment, I don't think I've got time to implement adaptive resampling from scratch. Would you also be interested in a fixed subsampling, whereby the user sets a number of samples in the UI, then, for example, the server subsamples at `TrackingServiceClient.get_metric_history`?

As far as I can tell, this would require changes at:
- A slider added to the `MetricsPlot` similar to existing the smoothness slider. Alternatively, a numeric text field may be sensible, whereby empty denotes all samples.
- `MetricsPlotPanel.loadMetricHistory` to add the value of the slider to the request
- `server.handlers._get_metric_history` to add the GET argument.
- `TrackingServiceClient.get_metric_history` to subsample
- Addition of a subsampling algorithm. Tensorboard uses [Reservoir sampling](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/reservoir.py) [wiki](https://en.wikipedia.org/wiki/Reservoir_sampling), but this seems to only make sense when we have an unknown number of samples, which we know at `TrackingServiceClient.get_metric_history`. The [plotly-resampler library](https://github.com/predict-idlab/plotly-resampler/blob/main/plotly_resampler/aggregation/aggregators.py) uses the [Largest-Triangle-Three-Buckets](https://pypi.org/project/lttbc/) algorithm ([demo](https://www.base.is/flot/)), which seems like a good choice. 

Does something like that seem reasonable to you? Am I missing any necessary changes?

**Comment by dbczumar** [comment]
> Hi @dbczumar, It seems like the library I mentioned is based on the Dash framework and thus appears difficult to integrate with the existing code-base. At the moment, I don't think I've got time to implement adaptive resampling from scratch. Would you also be interested in a fixed subsampling, whereby the user sets a number of samples in the UI, then, for example, the server subsamples at `TrackingServiceClient.get_metric_history`?
> 
> As far as I can tell, this would require changes at:
> 
> * A slider added to the `MetricsPlot` similar to existing the smoothness slider. Alternatively, a numeric text field may be sensible, whereby empty denotes all samples.
> * `MetricsPlotPanel.loadMetricHistory` to add the value of the slider to the request
> * `server.handlers._get_metric_history` to add the GET argument.
> * `TrackingServiceClient.get_metric_history` to subsample
> * Addition of a subsampling algorithm. Tensorboard uses [Reservoir sampling](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/reservoir.py) [wiki](https://en.wikipedia.org/wiki/Reservoir_sampling), but this seems to only make sense when we have an unknown number of samples, which we know at `TrackingServiceClient.get_metric_history`. The [plotly-resampler library](https://github.com/predict-idlab/plotly-resampler/blob/main/plotly_resampler/aggregation/aggregators.py) uses the [Largest-Triangle-Three-Buckets](https://pypi.org/project/lttbc/) algorithm ([demo](https://www.base.is/flot/)), which seems like a good choice.
> 
> Does something like that seem reasonable to you? Am I missing any necessary changes?

@pimdh Thank you for your suggestion! Unfortunately, it's challenging to make such changes to the MLflow service API without substantial debate & coordination because it's implemented by a variety of third-party providers (e.g. Databricks, AzureML). Since the issue appears to be restricted to Plotly performance, can we subsample on the frontend before passing datapoints to plotly?

**Comment by Fitree** [comment]
@dbczumar Is there any progress on this issue? Now, I'm suffering from this issue in my experiments.
Or could you suggest any way to bypass this issue if exists?

**Comment by Numeri** [comment]
I would love to see this feature as well! This (done on the server side) would reduce time to load and memory consumption considerably for pages with lots of runs/plots/data points.

**Comment by brarsukhAmazon** [comment]
Hi MLflow maintainers! 👋

I've investigated this 100k limit issue and found that **you've already solved the core problem** - but only for some components.

## What You've Already Built ✅

The sampling infrastructure is excellent and working:
- Backend APIs: `get_metric_history_bulk_interval` with `MAX_RESULTS_PER_RUN = 2500`
- Modern components: `RunsChartsLineChartCard` handles large datasets perfectly with greatly reduced client memory usage
- No need for Plotly Dash migration - your current React approach works great!

## The Gap 🔧

Legacy components like `MetricsPlotPanel` and `MetricsPlotView` still use the old unsampled APIs, causing high memory usage and the original client-side performance problems.

**Root cause:** These components are calling the old `/ajax-api/2.0/mlflow/metrics/get-history` endpoint instead of the new sampling endpoint `/ajax-api/2.0/mlflow/metrics/get-history-bulk-interval`.

## Simple Solution Request

Could you **connect the existing sampling APIs to the legacy MetricsPlotPanel**? 

The infrastructure is there - it just needs the same API integration that `RunsChartsLineChartCard` already uses. This would:
- Fix the memory usage issue completely
- Maintain consistent UX across all metric views  
- Leverage your existing, proven sampling solution

## Evidence

Timeline shows partial implementation:
1. **PR #7864**: Added "Too many metrics" UI limits (workaround)
2. **PR #11182**: Built sampling backend (solution)  
3. **PR #12584**: Added GraphQL endpoint (enhancement)
4. **Missing**: Connect legacy components to sampling APIs

The hard work is done - just needs the final connection! 🚀

---

*Analysis based on MLflow 3.1.1+ codebase investigation*

**Comment by brarsukhAmazon** [comment]
Following up on the above: for the change, should MetricsView be updated to:
1. Use the ChartCard directly?
2. Be refactored to React Hooks from Redux?

I've noticed some PRs coming in with some UI improvements, and would want to make sure the PR that would solve this issue aligns with the future direction

**Comment by daniellok-db** [comment]
@brarsukhAmazon thanks for this initiative! If it's possible to use the chart cards directly, that would be great. I'm not actually sure if using the redux hooks will work here (it's a different route altogether from the experiment runs page), but if it does then feel free to reuse that.

If not, then internally we're actually trying to move away from redux for new use-cases. We're finding it easier to work with individual reusable data fetching hooks (using `useQuery`, [example](https://github.com/mlflow/mlflow/blob/d48712345c5aaebfe9863ad892886a3abb8ab1e5/mlflow/server/js/src/experiment-tracking/hooks/logged-models/useGetLoggedModelQuery.tsx#L22-L50)).
